api/app/events/api.py

```diff
@@ -78,16 +80,173 @@ def _log_application_status(context):
+    except:
+        LOGGER.error("Encountered unknown error: {}".format(
+            traceback.format_exc()))
 
     _log_application_status('not available')
     return "Application not available"
 
 
+class EventAPI(EventMixin, restful.Resource):
+    @auth_required
+    def post(self):
+        args = self.req_parser.parse_args()
+
+        user_id = g.current_user["id"]
+        current_user = user_repository.get_by_id(user_id)
+        if not current_user.is_admin:
+            return FORBIDDEN
+
+        _date_format = '%Y-%m-%dT%H:%M:%S.%fZ'
+        name = args['name']
+        description = args['description']
+        start_date = datetime.strptime(
+            (args['start_date']), _date_format)
+        end_date = datetime.strptime(
+            (args['end_date']), _date_format)
+        key = args['key']
+        organisation_id = args['organisation_id']
+        email_from = args['email_from']
+        url = args['url']
+        application_open = datetime.strptime(
+            (args['application_open']), _date_format)
+        application_close = datetime.strptime(
+            (args['application_close']), _date_format)
+        review_open = datetime.strptime(
+            (args['review_open']), _date_format)
+        review_close = datetime.strptime(
+            (args['review_close']), _date_format)
+        selection_open = datetime.strptime(
+            (args['selection_open']), _date_format)
+        selection_close = datetime.strptime(
+            (args['selection_close']), _date_format)
+        offer_open = datetime.strptime(
+            (args['offer_open']), _date_format)
+        offer_close = datetime.strptime(
+            (args['offer_close']), _date_format)
+        registration_open = datetime.strptime(
+            (args['registration_open']), _date_format)
+        registration_close = datetime.strptime(
+            (args['registration_close']), _date_format)
+
+        event = Event(
+            name,
+            description,
+            start_date,
+            end_date,
+            key,
+            organisation_id,
+            email_from,
+            url,
+            application_open,
+            application_close,
+            review_open,
+            review_close,
+            selection_open,
+            selection_close,
+            offer_open,
+            offer_close,
+            registration_open,
+            registration_close
+        )
+
+        try:
+            event = event_repository.add(event)
+        except IntegrityError:
+            LOGGER.error("Event with KEY: {} already exists".format(key))
+            return EVENT_KEY_IN_USE
+
+        try:
+            admin_event_role = event_role_repository.add(
+                'admin', user_id, event.id)
+        except Exception as e:
+            LOGGER.error(
+                'Failed to add event role for user id {} due to: {}'.format(user_id, e))
+            return ADD_EVENT_ROLE_FAILED
```

Comment:
So I want to try to avoid this separate save to the db and rather try to save the event and event role at once. Can you try this? Setup the relationship between `Event` and `EventRole` by going to the `Event` model and adding a field called `event_roles = db.relationship('EventRole')` and in `EventRole` adding `event = db.relationship('Event'. foreign_keys=[event_id])`.

In the Event model, add the following method (you may need to initialise `event_roles` if it's `None` here or in the constructor)
```
def add_event_role(event_role: EventRole):
    self.event_roles.append(event_role)
```
Back in this file you can create your `EventRole`, and add it to the `Event` by running the `add_event_role()` method and then run `event_repository.add(event)`

Reply:
This change works.
Just a small adjustment I made was, the inputs to `add_event_role` are just the `role` and `user_id` 

```
def add_event_role(self, role, user_id):
    event_role = EventRole(role, user_id, self.id)
    self.event_roles.append(event_role)
```

Reply:
Oh yes of course. This is right. The event_id should just auto-populate (and yeah it doesn't event exist at this point)